cmd/insights/contributors.go

```diff
@@ -49,84 +52,132 @@ func NewContributorsCommand() *cobra.Command {
 		RunE: func(cmd *cobra.Command, args []string) error {
 			endpointURL, _ := cmd.Flags().GetString(constants.FlagNameEndpoint)
 			opts.APIClient = api.NewGoClient(endpointURL)
-			return run(opts)
+			output, _ := cmd.Flags().GetString(constants.FlagNameOutput)
+			opts.Output = output
+			return opts.run(context.TODO())
 		},
 	}
 	cmd.Flags().StringVarP(&opts.FilePath, constants.FlagNameFile, "f", "", "Path to yaml file containing an array of git repository urls")
 	cmd.Flags().Int32VarP(&opts.Period, constants.FlagNamePeriod, "p", 30, "Number of days, used for query filtering")
 	return cmd
 }
 
-func run(opts *Options) error {
+func (opts *contributorsOptions) run(ctx context.Context) error {
 	repositories, err := utils.HandleRepositoryValues(opts.Repos, opts.FilePath)
 	if err != nil {
 		return err
 	}
 	var (
-		waitGroup = new(sync.WaitGroup)
-		errorChan = make(chan error, len(repositories))
+		waitGroup    = new(sync.WaitGroup)
+		errorChan    = make(chan error, len(repositories))
+		insightsChan = make(chan contributorsInsights, len(repositories))
+		doneChan     = make(chan struct{})
+		insights     = make(contributorsInsightsSlice, 0, len(repositories))
+		allErrors    error
 	)
-	for url := range repositories {
-		repoURL := url
-		waitGroup.Add(1)
-		go func() {
-			defer waitGroup.Done()
-			repoContributorsInsights, err := findAllRepositoryContributorsInsights(context.TODO(), opts, repoURL)
+	go func() {
+		for url := range repositories {
+			repoURL := url
+			waitGroup.Add(1)
+			go func() {
+				defer waitGroup.Done()
+				allData, err := findAllContributorsInsights(ctx, opts, repoURL)
+				if err != nil {
+					errorChan <- err
+					return
+				}
+				if allData == nil {
+					return
+				}
+				insightsChan <- *allData
+			}()
+		}
+		waitGroup.Wait()
+		close(doneChan)
+	}()
+	for {
+		select {
+		case err = <-errorChan:
+			allErrors = errors.Join(allErrors, err)
+		case data := <-insightsChan:
+			insights = append(insights, data)
+		case <-doneChan:
+			if allErrors != nil {
+				return allErrors
+			}
+			output, err := insights.BuildOutput(opts.Output)
 			if err != nil {
-				errorChan <- err
-				return
+				return err
 			}
-			repoContributorsInsights.RenderTable()
-		}()
-	}
-	waitGroup.Wait()
-	close(errorChan)
-	var allErrors error
-	for err = range errorChan {
-		allErrors = errors.Join(allErrors, err)
+			fmt.Print(output)
```

Comment:
This is missing a newline at the end of the output which causes a `%` to be dropped at the end of output (indicating there's no newline before on my system ZSH drops in the prompt):

```
❯ pizza insights repositories https://github.com/open-sauced/app
Repository URL         https://github.com/open-sauced/app
Repository ID          501028599
All pull requests      73
Accepted pull requests 56
Spam pull requests     0
Contributors           19                                 %
❯ 
```

The quickfix is to just

```suggestion
			fmt.Println(output)
```